### PR TITLE
Allow WebGPU &  WebGL in same wasm and detect WebGPU availability

### LINKED ISSUE #2804: Dynamic WebGPU support detection
**Is your feature request related to a problem? Please describe.**
It's not really a problem at the moment, because no stable browser has WebGPU support turned on by default, yet. In the future, it would be great to be able to ship a single binary in WASM that supports both WebGL and WebGPU, depending on browser support.
Currently this is impossible, as selecting the `webgl` crate feature automatically disables the WebGPU backend.

**Describe the solution you'd like**
Detecting WebGPU support in the browser is fairly easy. So depending on it's availability, the correct backend should be chosen automatically by `wgpu`.

**Describe alternatives you've considered**
The only solution I can come up with right now, is to have two binaries. Either detect support in the browser and load the appropriate WASM binary, or ship the appropriate binary by the server depending on the requesting browser.

### PR DESCRIPTION
**Pull request by Wumpf**
**Connections**
* Based on #4984
* Fixes #2804

Shoutout to @daxpedda who did a lot of the groundwork to make this possible!

**Description**
Allows to compile with both webgpu & webgl backends enabled. There's a new `webgpu` feature flag that is enabled by default for this purpose. When additionally enabling `webgl` this will no longer disable `webgpu` and both features can live side by side.
The WebGPU backend remains a special case in that far that an instance decides (via very simple support detection logic) upon creation if it is a WebGPU instance or a WgpuCore instance. I did a bunch of renamings in that area to make this more obvious. Also, all `as_hal` methods now return an `Option` (most already did) to make it possible for them to return `None` whenever both WebGPU & WebGL are enabled (only a WebGL/WgpuCore instance can do hal conversion). Similar patter applies to a variety of things.

**DRAFT until #4984 lands**
(makes the diff also *significantly* smaller)

**Testing**
* clippy etc.
* local web tests to make sure they still do webgl when asked to (workarounds were needed, see code comments)
* there's an egui branch that works with this https://github.com/emilk/egui/tree/wgpu-update

image

**Checklist**

- [x] Run `cargo fmt`.
- [x] Run `cargo clippy`. If applicable, add:
  - [x] `--target wasm32-unknown-unknown`
  - [x] `--target wasm32-unknown-emscripten`
- [x] Run `cargo xtask test` to run tests.
- [x] Add change to `CHANGELOG.md`. See simple instructions inside file.

### CODE REVIEW

File: wgpu/src/backend/mod.rs [hunk 1]
```diff
@@ -1,9 +1,9 @@
...
 #[cfg(webgpu)]
-pub(crate) use web::Context;
+pub(crate) use webgpu::{get_browser_gpu_property, ContextWebGpu};
 
-#[cfg(not(webgpu))]
-mod direct;
-#[cfg(not(webgpu))]
-pub(crate) use direct::Context;
+#[cfg(wgpu_core)]
+mod wgpu_core;
```
Comment (cwfitzgerald) [inline wgpu/src/backend/mod.rs:7]: This isn't a great name as it's a conflict I wonder if there's a better option

File: wgpu/src/lib.rs [hunk 2]
```diff
@@ -1999,13 +2029,24 @@ impl Instance {
     }
 
     /// Generates memory report.
-    #[cfg(not(webgpu))]
+    ///
+    /// Like `try_generate_report` but panics if the active backend is WebGPU.
```
Comment (cwfitzgerald) [inline wgpu/src/lib.rs]: Couldn't this just be optional?

**Review by cwfitzgerald — CHANGES_REQUESTED** [review]
Need to see it rebased to fully review, but some initial comments

File: wgpu/src/backend/mod.rs (hunk 1 above)
Comment (Wumpf) [reply wgpu/src/backend/mod.rs:7]: maybe just `core`? I found the previous `direct` to be super confusing back in the day I figured this is a good opportunity to rename. Arguably we can also take this discussion/change separately :)

File: wgpu/src/lib.rs (hunk 2 above)
Comment (Wumpf) [reply wgpu/src/lib.rs]: tbh I have the two versions mostly I found it too tedious to patch all existing `generate_report` calls. Can revisit

File: wgpu/src/lib.rs [hunk 3]
```diff
@@ -1809,34 +1833,40 @@ impl Instance {
...
             context: Arc::new(unsafe {
-                crate::backend::Context::from_core_instance(core_instance)
+                crate::backend::ContextWgpuCore::from_core_instance(core_instance)
             }),
         }
     }
 
     /// Retrieves all available [`Adapter`]s that match the given [`Backends`].
     ///
+    /// Always returns an empty iterator if the instance decided upon creation to
```
Comment (Wumpf) [inline wgpu/src/lib.rs]:
```suggestion
    /// Always returns an empty vector if the instance decided upon creation to
```

would be nice to still return an iterator, but for that we'd need `itertools::Either`

File: wgpu/src/lib.rs [hunk 4]
```diff
@@ -2987,10 +3035,13 @@ impl<'a> BufferSlice<'a> {
...
-    /// which unconditionally copies mapped data into the wasm heap, this function directly hands you the
-    /// ArrayBuffer that we mapped the data into in js.
-    #[cfg(webgpu)]
+    /// This is useful when targeting WebGPU and you want to pass mapped data directly to js.
+    /// Unlike `get_mapped_range` which unconditionally copies mapped data into the wasm heap,
+    /// this function directly hands you the ArrayBuffer that we mapped the data into in js.
+    ///
+    /// If you are not targeting WebGPU, this function will perform an additional copy from the wasm heap to
+    /// the returned array buffer.
+    #[cfg(all(not(native), not(target_os = "emscripten")))]
```
Comment (Wumpf) [inline wgpu/src/lib.rs]:
```suggestion
    #[cfg(any(webgpu, webgl))]
```

File: wgpu/src/lib.rs (hunk 2 above)
Comment (Wumpf) [reply wgpu/src/lib.rs]: remove the extra report method again, wasn't _that_ bad

File: wgpu/src/backend/wgpu_core.rs [hunk 5]
```diff
@@ -1537,6 +1539,19 @@ impl crate::Context for Context {
...
+        &self,
+        buffer: &Self::BufferId,
+        buffer_data: &Self::BufferData,
+        sub_range: Range<wgt::BufferAddress>,
+    ) -> js_sys::ArrayBuffer {
+        let mapped_range = self.buffer_get_mapped_range(buffer, buffer_data, sub_range);
+        let array_buffer = js_sys::Uint8Array::new_with_length(mapped_range.slice().len() as u32);
+        array_buffer.copy_from(mapped_range.slice());
+        array_buffer.buffer()
+    }
```
Comment (cwfitzgerald) [inline wgpu/src/backend/wgpu_core.rs]: I don't think this is correct - the changes to the array buffer won't be reflected in the upstream changes.

**Review by cwfitzgerald — CHANGES_REQUESTED** [review]
Looks good outside of a issue with get_mapped_range

File: wgpu/src/backend/wgpu_core.rs (hunk 5 above)
Comment (Wumpf) [reply wgpu/src/backend/wgpu_core.rs]: ah you're right. We'd need a special unmap operation then! We'll have to make this function webgpu only and thus return an `Option` in the dynamic interface provided by `instance`

File: wgpu/src/backend/mod.rs (hunk 1 above)
Comment (Wumpf) [reply wgpu/src/backend/mod.rs:7]: changed your mind and just going with `wgpu_core` as module name despite the clash with the feature? 🤷 fine by me either way

File: wgpu/src/backend/wgpu_core.rs (hunk 5 above)
Comment (Wumpf) [reply wgpu/src/backend/wgpu_core.rs]: addressed in [6c10575](https://github.com/gfx-rs/wgpu/pull/5044/commits/6c105754e92da2997aeb15d461815aeaec94f2da)

**Review by cwfitzgerald — APPROVED** [review]
Hell yes!

**Comment by cwfitzgerald** [comment]
@Wumpf please resolve conflicts, then merge

**Comment by Wumpf** [comment]
woo 🥳

**Comment by Keavon** [comment]
This looks very handy! Is there a documented list of things the WebGL version doesn't support? I assume compute shaders is one such obvious missing feature (although I wonder if there's been any effort put towards polyfilling compute shaders with fragment shader magic?).

**Comment by Wumpf** [comment]
@Keavon there's unfortunately no written out list since it's a bit hard to capture - some of the exact capabilities are determined at runtime. But you can check https://docs.rs/wgpu/latest/wgpu/struct.DownlevelFlags.html# for possible lacking features as well as https://docs.rs/wgpu/latest/wgpu/struct.Limits.html#method.downlevel_webgl2_defaults which lists out the reduced limits of webgl directly.
As you guessed there's no compute shaders and no effort to polyfill them - this would be huge undertaking and likely run into issues that would make you resort to cpu based emulation (most prominently there's no random access writes at all which bars the door to emulating the harder stuff like shared memory and don't even get me started about atomics ;))

**Comment by cwfitzgerald** [comment]
It's almost certainly easier to polyfill them _in the user's application_ using application specific knowledge if you need them.

**Pull request merged**